File: ios/stream/stream/option.swift
//
//  option.swift
//  stream
//
//  Created by xpwu on 2021/3/24.
//

import Foundation

public struct Config {
  var host:String
  var port:Int
  var tls:Bool
  var connectTimeout:Duration
  var requestTimeout:Duration
}

public typealias Option = (inout Config)->Void

public func Host(_ host:String) -> Option {
  return {(config: inout Config)->Void in
    config.host = host
  }
}

public func Port(_ port:Int) -> Option {
  return {(config: inout Config)->Void in
    config.port = port
  }
}

public func TLS(_ tls:Bool) -> Option {
  return {(config: inout Config)->Void in
    config.tls = tls
  }
}

public func ConnectTimeout(_ duration: Duration)-> Option {
  return {(config: inout Config)->Void in
    config.connectTimeout = duration
  }
}

public func RequestTimeout(_ duration: Duration)-> Option {
  return {(config: inout Config)->Void in
    config.requestTimeout = duration
  }
}

File: android/app/src/main/java/com/anywithyou/stream/option.kt
@file:JvmName("Option_Kt")

package com.anywithyou.stream

import java.net.Socket

class OptionKt (internal val o : Option)

fun Host(host: String): OptionKt {
	return OptionKt(Option.Host(host))
}

fun Port(port: Int): OptionKt {
	return OptionKt(Option.Port(port))
}

fun TLS(): OptionKt {
	return OptionKt(Option.TLS())
}

fun TLS(strategy: (host: String, port: Int , tcpSocket: Socket)->Socket): OptionKt {
	return OptionKt(Option.TLS(strategy))
}

fun ConnectTimeout(duration: DurationKt): OptionKt {
	return OptionKt(Option.ConnectTimeout(Duration(duration.d)))
}

fun RequestTimeout(duration: DurationKt): OptionKt {
	return OptionKt(Option.RequestTimeout(Duration(duration.d)))
}

internal fun Array<out OptionKt>.toOptions(): Array<Option> {
	val l = this.map { return@map it.o }
	return l.toTypedArray()
}